src/main/java/hexgui/sgf/SgfWriter.java

```diff
@@ -0,0 +1,161 @@
+// ----------------------------------------------------------------------------
+// $Id$
+// ----------------------------------------------------------------------------
+
+package hexgui.sgf;
+
+import hexgui.game.GameInfo;
+import hexgui.game.Node;
+import hexgui.hex.HexColor;
+import hexgui.hex.HexPoint;
+import hexgui.hex.Move;
+import hexgui.version.Version;
+import java.awt.Dimension;
+import java.io.*;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Vector;
+
+// ----------------------------------------------------------------------------
+
+/** SGF Writer. See https://www.red-bean.com/sgf/ for the SGF definition. */
+public final class SgfWriter {
+
+  /** Write a game tree. */
+  public SgfWriter(OutputStream out, Node root, GameInfo game) {
+    m_out = new PrintStream(out);
+    m_buffer = new StringBuffer(128);
+    m_gameinfo = game;
+
+    writeTree(root, true);
+    print("\n");
+    flushBuffer();
+    m_out.flush();
+    m_out.close();
+  }
+
+  private void writeTree(Node root, boolean isroot) {
+    print("(");
+    writeNode(root, isroot);
+    print(")");
+  }
+
+  private void writeNode(Node node, boolean isroot) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 43-57, lines 63-77, lines 79-96, lines 98-108 into separate methods.